presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.zorder;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.String.format;
+
+/**
+ * This class contains functions to search for addresses within given value ranges.
+ */
+public class ZOrderCurveSearcher
+{
+    private final int[] dimensions;
+
+    private final ZOrder zOrder;
+
+    private List<ZValueRange> ranges;
+
+    private boolean returnLong;
+    private final LinkedList<ZAddressRange> addressRanges;
+
+    /**
+     * Class constructor specifying the ZOrder class and initializing <code>dimensions</code>.
+     * <p/>
+     * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+     */
+    ZOrderCurveSearcher(ZOrder zOrder)
+    {
+        this.zOrder = zOrder;
+
+        this.addressRanges = new LinkedList<>();
+
+        this.dimensions = new int[zOrder.maxBitLength];
+        initializeDimensions();
+    }
+
+    private void initializeDimensions()
+    {
+        List<Integer> bitLengths = new ArrayList<>(zOrder.encodingBits);
+        for (int dimensionIndex = zOrder.maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+            for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                if (bitLengths.get(bitLengthIndex) > 0) {
+                    dimensions[dimensionIndex]++;
+                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                }
+            }
+        }
+    }
+
+    void setReturnLong(boolean returnLong)
+    {
+        this.returnLong = returnLong;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    List<ZAddressRange> zOrderCondensedSearchCurve(List<ZValueRange> ranges)
+            throws Exception
```

Comment:
What are the exceptions that the caller is expected to handle? If it is runtime exception, let's not declare in the method. If not, let's be concrete about the checked exception. `Exception` is too abstract.